root.c

```diff
@@ -1922,6 +1938,161 @@ static void notify_thread(struct unlocked_watchman_root *unlocked)
+    } else {
+      res = find_project_root(root_files, resolved, &relpath);
+      if (!res) {
+        w_log(W_LOG_ERR, "No root project found to contain %s\n", resolved);
+      } else {
+        struct unlocked_watchman_root unlocked;
+        bool success = w_root_resolve(resolved, true, &errmsg, &unlocked);
+        if (!success) {
+          w_log(W_LOG_ERR, "watch_symlink_target: failed to watch %s\n",
+                resolved);
+        }
+      }
+      json_decref(root_files);
+    }
+    // Freeing resolved also frees rel_path
+    free(resolved);
+  }
+  w_string_delref(normalized_target);
+}
+
+/** Given an absolute path, watch all symbolic links associated with the path.
+ * Since the target of a symbolic link might contain several components that
+ * are themselves symlinks, this function gets called recursively called on
+ * all the components of path. */
+static void watch_symlinks(w_string_t *path) {
+  w_string_t *dir_name, *file_name;
+  char link_target_path[WATCHMAN_NAME_MAX];
+  ssize_t tlen = 0;
+
+  // We do not currently support symlinks on Windows, so comparing path to "/"
+  // is ok
+  if (!path || w_string_equal_cstring(path, "/")) {
+    return;
+  }
+  w_assert(w_is_path_absolute(path->buf),
+           "watch_symlinks: path %s is not absolute\n", path->buf);
+  dir_name = w_string_dirname(path);
+  file_name = w_string_basename(path);
+  tlen = readlink(path->buf, link_target_path, WATCHMAN_NAME_MAX);
+  if (tlen >= WATCHMAN_NAME_MAX) {
+    w_log(W_LOG_ERR,
+          "readlink(%s), symlink target is too long: %d chars >= %d chars\n",
+          path->buf, (int)tlen, WATCHMAN_NAME_MAX);
+  } else if (tlen < 0) {
+    if (errno == EINVAL) {
+      // The final component of path is not a symbolic link, but other
+      // components in the path might be symbolic links
+      watch_symlinks(dir_name);
+    } else {
+      w_log(W_LOG_ERR,
+          "readlink(%s) errno=%d tlen=%d\n", path->buf, errno, (int)tlen);
+    }
+  } else {
+    w_string_t *target = w_string_new_len_typed(
+        link_target_path, tlen, W_STRING_BYTE);
+    if (w_is_path_absolute(target->buf)) {
+      watch_symlink_target(target);
+      watch_symlinks(target);
+      watch_symlinks(dir_name);
+    } else {
+      w_string_t *absolute_target = w_string_path_cat(dir_name, target);
+      watch_symlink_target(absolute_target);
+      watch_symlinks(absolute_target);
+      // No need to watch_symlinks(dir_name), since
+      // watch_symlinks(absolute_target) will eventually have the same effect
+      w_string_delref(absolute_target);
+    }
+    w_string_delref(target);
+  }
+  if (dir_name) {
+    w_string_delref(dir_name);
+  }
+  if (file_name) {
+    w_string_delref(file_name);
+  }
+}
+
+/** Process the list of observed changed symlinks and arrange to establish
+ * watches for their new targets */
+static void process_pending_symlink_targets(w_root_t *root) {
```

Comment:
I suggested elsewhere that you pass in the unlocked pointer here.
It's also worth noting in a comment that it is safe for us to operate on `unlocked->root->pending_symlink_targets` because we only ever mutate it from the IO thread.